anndata/_io/merge.py

```diff
@@ -0,0 +1,647 @@
+        }
+    )
+    for k in keys:
+        elems = [m[k] for m in mappings]
+        write_concat_sequence(
+            elems,
+            output_group=mapping_group,
+            out_path=k,
+            axis=axis,
+            index=index,
+            reindexers=reindexers,
+            fill_value=fill_value,
+        )
+
+
+SPARSE_MATRIX = {"csc_matrix", "csr_matrix"}
+
+EAGER_TYPES = {"dataframe", "awkward-array"}
+
+
+def read_group(group: Union[ZarrGroup, H5Group]):
+    """Read the group until
+    SparseDataset, Array or EAGER_TYPES are encountered."""
+
+    def callback(func, elem_name: str, elem, iospec):
+        if iospec.encoding_type in SPARSE_MATRIX:
+            return SparseDataset(elem)
+        elif iospec.encoding_type in EAGER_TYPES:
+            return read_elem(elem)
+        elif iospec.encoding_type == "array":
+            return elem
+        else:
+            return func(elem)
+
+    return read_dispatched(group, callback=callback)
+
+
+def read_only_dict(group: Union[ZarrGroup, H5Group]):
+    """Read the dict and leave the rest of the group alone."""
+
+    def callback(func, elem_name: str, elem, iospec):
+        if iospec.encoding_type == "dict":
+            return {k: v for k, v in elem.items()}
+        else:
+            return elem
+
+    return read_dispatched(group, callback=callback)
+
+
+def get_encoding_type(group: Union[ZarrGroup, H5Group]) -> str:
+    """Get the encoding type of a group."""
+    return group.attrs.get("encoding-type")
+
+
+def get_shape(group: Union[ZarrGroup, H5Group]) -> str:
+    """Get the shape of a group (array, sparse matrix, etc.)"""
+    if group.attrs["encoding-type"] == "array":
+        return group.shape
+    elif group.attrs["encoding-type"] in SPARSE_MATRIX:
+        return group.attrs.get("shape")
+    raise NotImplementedError(f"Cannot get shape of {group.attrs['encoding-type']}")
+
+
+def _gen_elem_to_append(elems, axis=0, reindexers=None, fill_value=None):
+    for elem, ri in zip(elems, reindexers):
+        if ri.no_change:
+            yield elem
+        else:
+            yield ri(elem, axis=1 - axis, fill_value=fill_value)
+
+
+def gen_reindexers_df_inner(dfs: Sequence[pd.DataFrame], axis: Literal[0, 1] = 0):
+    if axis == 0:
+
+        def df_indices(x):
+            return x.columns
+
+    elif axis == 1:
+
+        def df_indices(x):
+            return x.indices
+
+    # TODO Handle empty later
+    common_ind = reduce(lambda x, y: x.intersection(y), (df_indices(df) for df in dfs))
+    reindexers = [Reindexer(df_indices(df), common_ind) for df in dfs]
+    return reindexers
+
+
+def gen_reindexers_array_inner(
+    arrays: Sequence[Union[ZarrArray, H5Array, SparseDataset]], axis: Literal[0, 1] = 0
+):
+    min_ind = min(a.shape[1 - axis] for a in arrays)
+    reindexers = [
+        gen_reindexer(pd.RangeIndex(min_ind), pd.RangeIndex(a.shape[1 - axis]))
+        for a in arrays
+    ]
+
+    return reindexers
```

Comment:
I think this might be duplicated from the in-memory concat. Let me check in on that, and see if there's an easy way to use those methods.

Reply:
true, solved it